mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMailboxFlagInconsistenciesTaskDTO.java

```diff
@@ -0,0 +1,67 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.cassandra.mail.task;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.james.json.DTOModule;
+import org.apache.james.mailbox.cassandra.mail.task.SolveMailboxFlagInconsistencyTask.TargetFlag;
+import org.apache.james.server.task.json.dto.TaskDTO;
+import org.apache.james.server.task.json.dto.TaskDTOModule;
+import org.apache.james.task.TaskType;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+
+public record SolveMailboxFlagInconsistenciesTaskDTO(@JsonProperty("type") String type,
+                                                     @JsonProperty("flagName") String flagName) implements TaskDTO {
+
+    public static final String TYPE = "solve-mailbox-flag-inconsistencies";
+    public static final TaskType TASK_TYPE = TaskType.of(TYPE);
+
+    public static TaskDTOModule<SolveMailboxFlagInconsistencyTask, SolveMailboxFlagInconsistenciesTaskDTO> module(
+        SolveMailboxFlagInconsistenciesService mailboxDeletedFlagResolveService,
+        SolveMailboxFlagInconsistenciesService mailboxRecentFlagResolveService) {
+        return DTOModule
+            .forDomainObject(SolveMailboxFlagInconsistencyTask.class)
+            .convertToDTO(SolveMailboxFlagInconsistenciesTaskDTO.class)
+            .toDomainObjectConverter(getDomainObjectConverter(mailboxDeletedFlagResolveService, mailboxRecentFlagResolveService))
+            .toDTOConverter((task, type) -> new SolveMailboxFlagInconsistenciesTaskDTO(type, task.targetFlag()))
+            .typeName(TYPE)
+            .withFactory(TaskDTOModule::new);
+    }
+
+    private static DTOModule.DomainObjectConverter<SolveMailboxFlagInconsistencyTask, SolveMailboxFlagInconsistenciesTaskDTO> getDomainObjectConverter(SolveMailboxFlagInconsistenciesService mailboxDeletedFlagResolveService,
+                                                                                                                                                       SolveMailboxFlagInconsistenciesService mailboxRecentFlagResolveService) {
+        return dto -> {
+            Preconditions.checkArgument(StringUtils.isNotEmpty(dto.flagName()), "flagName should not be empty");
+            return switch (TargetFlag.valueOf(dto.flagName())) {
+                case TargetFlag.DELETED ->
+                    new SolveMailboxFlagInconsistencyTask(mailboxDeletedFlagResolveService, TargetFlag.DELETED);
+                case TargetFlag.RECENT ->
+                    new SolveMailboxFlagInconsistencyTask(mailboxRecentFlagResolveService, TargetFlag.RECENT);
+            };
```

Comment:
Please do not put logic in deserialisation layer.

How about passing both mailboxDeletedFlagResolveService and mailboxRecentFlagResolveService as parameters of SolveMailboxFlagInconsistencyTask and letting it chose the right one based on the TargetFlag?

Actually that choice can even be moved into SolveMailboxFlagInconsistenciesService wich could remain injected...

Reply:
updated